libminifi/include/utils/ScopeGuard.h

```diff
@@ -0,0 +1,64 @@
+/**
+ * @file ScopeGuard.h
+ * Exception class declaration
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __SCOPE_GUARD_H__
+#define __SCOPE_GUARD_H__
+
+#include <functional>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+
+class ScopeGuard {
+ private:
+  bool enabled_;
+  std::function<void()> func_;
+
+ public:
+  ScopeGuard(std::function<void()>&& func)
+    : enabled_(true)
+    , func_(std::move(func)) {
+  }
+
+  ~ScopeGuard() {
+    if (enabled_) {
+      try {
+        func_();
```

Comment:
I loved scoped guards when used in C++ but you're calling the closure of c code, have you validated that the close functions on libssh don't trap and shut the agent down if a closure were to happen after some errant transfer ? It seems like those closures would be safe, but want to double check, thanks!

Reply:
I love them too.
libssh2_sftp_close should be called on every handle even if there were transfer errors, otherwise resources wouldn't be freed.
I assume you meant the closure question somewhat independently from using ScopeGuards though - if they would trap they would do so even if the function wouldn't be called from ~ScopeGuard.